[QPPA-606] uuid and strata extraction from ecqm xml files

kencheeto opened the pull request:
This PR adds two scripts to help integrate necessary strata and UUID details for eCQM measures from a zip of eCQM specs for [QPPA-606](https://jira.cms.gov/browse/QPPA-606).

The new data being added for eCQMs:
* measure UUIDs are programmatically extracted from zips
  > with the exception of CMS145v5 and CMS160v5, which were extracted manually by me since their data format isn't consistent with the others
* strata descriptions and strata UUIDs are extracted from the zips
* strata names are written by me, hopefully to the effect of summarizing their descriptions
* metricType is deduced to either `singlePerformanceRate` or `multiPerformanceRate` depending on if they have one or multiple strata
* overallAlgorithm is cross-referenced from a perf rate xls file linked in the JIRA issue

Generally, the new scripts work like this:
#### `node scripts/get-strata-and-uuids-from-ecqm-zip.js <path to zip>`
1. unzips the file containing the specifications for all eCQM measures
1. parses the `CMS<measure ID>v<measure version>_SimpleXML.xml` files within to extract strata descriptions, UUIDs, and metricType
1. extracted data is written to `utils/generated-ecqm-data.json`, keyed on eMeasureId (CMS137v5)

unfortunately, strata names and overallAlgorithm are not programmatically extractable from these source files, so we need another script. This one is only used once to build a scaffold for easily manually adding missing fields by hand, but is included for reuse should we need to:
#### `node scripts/find-ecqms-with-missing-data.js <path to measures-data.json>`
1. looks in measures-data.json to find eCQM measures that are missing strata names (these are also all missing overallAlgorithm)
1. builds a json file similar to measures-data.json that only includes the missing fields (as empty strings), along with strata descriptions (since these are useful for generating the strata names)
1. the two outlier eCQMs (CMS145v5 and CMS160v5) are extracted by hand from source and hardcoded into the json
1. the json file is written to `utils/manually-added-ecqm-data.json`
1. this file needs to be edited manually to add strata names based off of descriptions and overallAlgorithm based on another source (xref'd from an xls file)

third, merge-measures-data.js has been updated to merge in all of the new generated and manual eCQM data, preserving some fields where data is already present (some eCQM strata already had names).

## Pretty messy stuff - I want to call out a few things specifically for review:

1. [strata names](https://github.com/CMSgov/qpp-measures-data/pull/32/commits/ac73a35fafb8692d1f24bf265bf19d45a6f9e0d2): I wrote these based on descriptions, need someone to read through and sanity check that they aren't misleading
1. [manually extracted eCQMs](https://github.com/CMSgov/qpp-measures-data/pull/32/commits/f501ebc4e651b9dd66d5ae1db47e08dc51bfb552): The UUIDs are correct but the strata names _and_ descriptions for 145 & 160 are handwritten based on source files from the zip. These felt a little shaky on the wording
1. our general approach for assembling measures-data.json. See changes to the merge-measures-data.js script - merging in [generated](https://github.com/CMSgov/qpp-measures-data/pull/32/commits/fc5be5b3bf1ba754ead3efbc821ee4a938766cab) and [manually added](https://github.com/CMSgov/qpp-measures-data/pull/32/commits/62195d9dd5d62c5bccb4c3a465229806dd4f75ca) data. The current merge-measures-data approach is order-dependent and each source is treated inconsistently (I'm not helping here). Relatedly, we also need to cut our dependency on the measures API which is still used to pull in the bulk of measures - if this repo is the source of truth then we should anoint a file in this repo as the measures origin. I think there's some zooming out and refactoring to do.

## Things that don't need review:

1. the 2 new scripts - feel free to take a look, but I need to refactor to get from working to maintainable
1. the measures-data.json file, this was validated already

todos:
* [x] unzip nested xmls
* [x] extract data from xmls
* [x] verify strata UUID ordering against source data
* [x] update merge-measures-data.js to integrate extracted and manually added data
* [x] manually add CMS145v5 and CMS160v5 💦 
* [x] manually add strata names 💦 
* [x] regenerate measures-data.xml
* [x] add ecqm fields to measures-schema.yaml
* [x] refactor pass

bijujoseph commented:
@marimiyachi , @kencheeto  - this change has no impact on QS and MA code right ?

kencheeto commented:
@bijujoseph eCQM measures have been updated with new strata (and corresponding changes to metricType, overallAlgorithm) - there should be no changes required for QS/MA code (no new fields), but some tests might need to be updated on your end.

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+    "eMeasureId": "CMS123v5",
+    "strata": [
+      {
+        "name": "receivedExams",
```
Is it worth being more specific here? Maybe `receivedFootExams` or even `receivedVisualPulseAndSensoryFootExams`?

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+    "eMeasureId": "CMS137v5",
+    "strata": [
+      {
+        "name": "treatmentWithin14Days",
```
Maybe `treatmentWithin14DaysOfDiagnosis`?

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+        "description": "Patients who initiated treatment within 14 days of the diagnosis"
+      },
+      {
+        "name": "treatmentPlusServicesWithin30",
```
Might be worth adding "days" to the end of it? Ie `treatmentPlusServicesWithin30DaysOfVisit`?

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+    "eMeasureId": "CMS155v5",
+    "strata": [
+      {
+        "name": "BMI",
```
Maybe `recordedHeightWeightBMI`?

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+    "eMeasureId": "CMS158v5",
+    "strata": [
+      {
+        "name": "hepBTested",
```
`hepatitisBTested` just for clarity?

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+    "eMeasureId": "CMS166v6",
+    "strata": [
+      {
+        "name": "noImagingStudy",
```
`noImagingStudyWithin28Days`?

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+    "eMeasureId": "CMS167v5",
+    "strata": [
+      {
+        "name": "dilatedMacularOrFundusExam",
```
`dilatedMacularOrFundusExamWithin12Months`?

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+    "eMeasureId": "CMS169v5",
+    "strata": [
+      {
+        "name": "alcoholAssessmentEvidence",
```
Might be misleading to just have `alcohol` in the name since the description says "alcohol or other substance use"? Maybe `alcoholOrOtherSubstanceAssessmentEvidence`?

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+    "eMeasureId": "CMS52v5",
+    "strata": [
+      {
+        "name": "within3MonthsCD4<200/mm3",
```
Maybe prefix with `prescribed`?

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+        "description": "Patients who were prescribed Pneumocystis jiroveci pneumonia (PCP) prophylaxis within 3 months of CD4 count below 200 cells/mm3"
+      },
+      {
+        "name": "within3MonthsCD4<500/mm3OrCD4<15%",
```
Maybe prefix with `prescribed`?

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+    "eMeasureId": "CMS65v6",
+    "strata": [
+      {
+        "name": "lowerFollowupBP",
```
`lowerFollowupBloodPressure`?

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+    "eMeasureId": "CMS136v6",
+    "strata": [
+      {
+        "name": "hadPractitionerVisit",
```
`hadPractitionerVisitWithin30Days`?

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+    "eMeasureId": "CMS149v5",
+    "strata": [
+      {
+        "name": "assessedCognition",
```
`assessedCognitionOnceWithin12Months`?

samskeller commented on util/manually-added-ecqm-data.json:
```diff
@@ -0,0 +1,329 @@
+    "metricType": "multiPerformanceRate",
+    "strata": [
+      {
+        "description": "Beta Blocker Therapy for LVSD",
```
I might be very off-base about this, but it looks like from the description between this stratum and the next that the difference is just about the therapy being "for LVSD" vs not, right? Looking at the "CMS145v4.html" file in the zip, though, I also see differences in the two "Population Criteria" in the "Denominator" sections... Should those differences be reflected in these descriptions? And the general description near the top of the file mentions patients who have coronary heart disease -- should that be mentioned in these descriptions too? (Again, might be very wrong here, I still don't really understand what this all means)

kencheeto replied:
Yeah, my descriptions were way too slim. I took a deeper look and produced these:

> Patients with at least 2 encounters with eligible professionals and left ventricular systolic dysfunction (LVEF <40%) who are prescribed beta-blocker therapy for LVSD. For these patients, beta-blocker therapy includes the following: bisoprolol, carvedilol, or sustained release metoprolol succinate.

and 

> Patients with at least 2 encounters with eligible professionals with a prior (resolved within last 3 years) myocardial infarction who are prescribed beta-blocker therapy. For patients with prior MI, beta-blocker therapy includes any agent within the beta-blocker drug class.

kencheeto commented:
thanks @samskeller ! updated

marimiyachi commented on measures/measures-schema.yaml:
```diff
@@ -82,6 +82,9 @@ definitions:
       eMeasureId:
         description: Identifier for Electronic Clinical Quality Measures (ECQM).
         type: ['null', string]
+      eMeasureUuid:
```
Are these required fields now for quality measures?

marimiyachi replied:
Looks like no, but they are also not nullable.

kencheeto replied:
Ahhh, I misunderstood how fields are optional unless defined explicitly - will make this a string-only type and leave it optional.

marimiyachi commented on scripts/find-ecqms-with-missing-data.js:
```diff
@@ -0,0 +1,119 @@
+
+const missingJson = outlierEcqms.concat(ecqmsWithMissingData);
+const sortedMissingJson = _.sortBy(missingJson, ['eMeasureId']);
+fs.writeFileSync(path.join(__dirname, '../util/manually-added-ecqm-data.json'), JSON.stringify(sortedMissingJson, null, 2));
```
Maybe write to standard out instead? Scripts that use standard in/out are little more flexible.

kencheeto replied:
agree! i will do that in the refactor pass.

marimiyachi replied:
Cool - maybe add a TODO?

kencheeto replied:
fixed!

marimiyachi commented on scripts/merge-measures-data.js:
```diff
@@ -47,9 +47,70 @@ function mergeQpp(qppJson) {
+        }
+      });
+      if (oldStrata.length > 1) {
+        // ASSUMPTION: strata already available are in the same order as the ones extracted from the xml files
```
😳

kencheeto replied:
The ordering is derived from the spec, so as long as we are generally always pulling things in the correct order from the spec we should be okay. Still, yeah, kinda 😬

marimiyachi commented on scripts/find-ecqms-with-missing-data.js:
```diff
@@ -0,0 +1,119 @@
+  .filter(measure => measure.strata.some(stratum => !stratum.name))
+  .map(measure => {
+    const strata = measure.strata
+      .filter(stratum => !stratum.name)
```
Why do you need to filter again here? Are there some measures that only one of the stratum's have names?

kencheeto replied:
Yeah, some eCQMs have one named stratum, so need to filter those out.

kencheeto replied:
(I'll add a comment to explain this)

marimiyachi replied:
Hm....It's a bit of a strange circular dependency going on here where we use `measures-data.json` to generate `manually-added-ecqm-data.json` and then use `manually-added-ecqm-data.json` to generate `measures-data.json`. Are you sure that this step is necessary? Presumably what is in `measures-data.json` pre this PR about eCQMs is invalid and after this PR it should be corrected (i.e. all stratums should have names).

kencheeto replied:
this script isn't meant to be run more than once unless new ecqms are added - maybe I shouldn't include it at all? it just generates an empty scaffold that can easily be edited by hand.

kencheeto replied:
i added clarifying instructions to this script, and it's not referenced in the readme, so I believe it shouldn't be confusing anymore: https://github.com/CMSgov/qpp-measures-data/pull/32/commits/758025dd5f326935926a5918ca6c96f5356f9234

marimiyachi commented:
Also GREAT JOB ON THIS. Amazing comments - really easy to follow along what is actually a super complicated process.

marimiyachi commented on scripts/merge-measures-data.js:
```diff
@@ -47,9 +47,70 @@ function mergeQpp(qppJson) {
+
+  // separate passes for ecqm data to keep things simple
+  // load extracted ecqm data
+  const generatedEcqmStrataJson = JSON.parse(fs.readFileSync(path.join(__dirname, '../util/generated-ecqm-data.json'), 'utf8'));
```
Does it make sense to handle this inline with the other performance rate data file processing (line 30)? Right now the `measuresNotFound` error is compiling measures that are likely represented here.

marimiyachi replied:
That way the override logic for individual fields could be more explicit?

kencheeto replied:
Yeah, I started that way, but wanted to keep them separate at first since we could reason about the new additions separately without also considering the original logic in merge-measures-data. For example, I ran into an issue with the guard clause exiting early when no perf rate info was found - I didn't want to change the resulting logic to consider when ecqm data was found but not perf rate data (and vice versa).

I'm thinking that the refactor can explicitly enumerate the different sources of data and the fields pulled from them.

marimiyachi replied:
Hm....ok I hear ya. The logging for `measuresNotFound` should be adjusted at the very least.

kencheeto replied:
I didn't find anything to add to `measuresNotFound` - I've already added logging for potential issues when adding ecqm data, so I moved the `measuresNotFound` flushing to the bottom to make it clear that it happens at the end.

kencheeto replied:
i added some comments to merge-measures-data to make the ecqm additions clearer: https://github.com/CMSgov/qpp-measures-data/pull/32/commits/b3fb1243c5ac53e040e9b2de5b320db2b60e3705

marimiyachi commented:
Last comment - in general with the strata names I'd say lean on the side of being as concise as possible. It's meant to be a differentiator between the other strata, not to provide all description.

kencheeto commented:
thanks @marimiyachi 🙏 ! I think I've addressed your comments (shortened strata names in https://github.com/CMSgov/qpp-measures-data/pull/32/commits/b7bab5fc4d2830647b1ec35203c57ed3af0bb146), this should be good to merge now! I wrote a followup issue [QPPA-687](https://jira.cms.gov/browse/QPPA-687) for refactoring the scripts. As-is, this PR incrementally worsens the complexity of the measures-data.json build process, but should be well documented enough to allow debugging/maintenance.

QPPA-687 will take a step back and includes suggestions for refactoring the build process.
